doc/man3/EVP_PKEY_size.pod

```diff
@@ -0,0 +1,40 @@
+=pod
+
+=head1 NAME
+
+EVP_PKEY_size, EVP_PKEY_bits, EVP_PKEY_security_bits
+- EVP_PKEY information functions
+
+=head1 SYNOPSIS
+
+ #include <openssl/evp.h>
+
+ int EVP_PKEY_size(const EVP_PKEY *pkey);
+ int EVP_PKEY_bits(const EVP_PKEY *pkey);
+ int EVP_PKEY_security_bits(const EVP_PKEY *pkey);
+
+=head1 DESCRIPTION
+
+EVP_PKEY_size() and EVP_PKEY_bits() returns the cryptographic length
```

Comment:
I believe that the saner thing to do is to deprecate EVP_PKEY_size as it is horribly named and the cause of this mess.

If, as the current internal implementations suggest, the return value of this function is to be used to preallocate buffers for the result of some specific operation on the key, it is best asked to the operation in question to avoid ambiguity on what length to return when a key supports more than just signatures for example.

Currently anyway we can already call the low level sign or derive operations with a NULL buffer pointer to retrieve the required buffer length to safely store the result of the operation.

Reply:
I can see the need for "give me an estimate of a signature size" or "give me an estimate of a shared secret derived by key exchange" from a given key without having to setup all the necessary context.

But again it is a signature over key or kexchange over key property, not a key property, so I would favor deprecation of EVP_PKEY_size and introduction of EVP_PKEY_{signature, exchange, encrypt, ... }_output_size.

We can then document that for legacy compatibility EVP_PKEY_size returns the first non-zero result of the operation specific methods (in a given order: signature, exchange, ...) or 0 if the key does not support any operation (useful key!)

Reply:
So, er, just woke up and reworded the description again.  Hold on.

Reply:
(for people reading the comments through email, I apologize but I had to edit a few times the comments above, they are all minor edits but for the last comment, where I also added more details and reworded more significantly)

Reply:
The name is, as you said, not the best choice.  However, if we deprecate it, we need to invent a new name, and for the life of me, I don't have enough imagination for the moment.

I suggest that we post-pone any renaming for another PR.  It might be that we get the manual to a point where we're both happy for now :smile:

Reply:
> The name is, as you said, not the best choice. However, if we deprecate it, we need to invent a new name, and for the life of me, I don't have enough imagination for the moment.

What about [this proposal from the edited comment above](https://github.com/openssl/openssl/pull/10778#discussion_r364626127)?



> But again it is a signature-over-key or kexchange-over-key property, not a key property, so I would favor deprecation of `EVP_PKEY_size` and introduction of `EVP_PKEY_{signature, kexchange, encrypt, ... }_output_size` (with the caveat that these abstract sizes will likely be just informative estimates and deferring to the actual `EVP_PKEY_<operation>` function documentation for the actual requirements on buffer sizes)
> 
> We can then document that for legacy compatibility `EVP_PKEY_size` returns the first non-zero result of the operation specific methods (in a given order: signature, exchange, ...) or 0 if the key does not support any operation (useful key!)



> I suggest that we post-pone any renaming for another PR. It might be that we get the manual to a point where we're both happy for now

It's ok to discuss this as a separate dedicated PR, but I would prefer this PR to be on hold until this issue is solved, as I believe conceptually this should either split into properties belonging to the `OSSL_DISPATCH` of each <operation> or if for some reason I am not seeing right now we want this to stay into the `keymgmt` it should be a

~~~C
OSSL_CORE_MAKE_FUNC(const size_t, OP_keymgmt_query_operation_size, (int operation_id))
~~~

Reply:
Er...  this PR already has the mechanism for getting the number for `EVP_PKEY_size()`...  that was the primary reason I submitted it!  Documentation was added later.

Reply:
Regarding new names, and apparently completely new ways to get the size for signatures, public key encrypted secret keys and whatnot, we still need to maintain `EVP_PKEY_size()`, and since that only gets an `EVP_PKEY` as input, the only recourse we have is to go through the KEYMGMT interface.

The suggested new names are tied to contexts for the diverse operations, and would of course mean going through each operation interface, and considering how that will in all likelyhood explode into quite an effort, I'd like to *strongly* suggest another PR where that path can be explored.

Reply:
See https://github.com/openssl/openssl/pull/10778#discussion_r364660278

I don't believe this PR could be considered complete if there is no clear and unambiguous indication of what should be the return for a query for `OSSL_PKEY_PARAM_SIZE`.

A `keymgmt` can be associated with more than one operation (e.g. both signature and kexchange like for EC) and right now we are not defining what to do in that case!